Update the section on using Hypertables (#231)
timescale/docs.timescale.com-content: using-timescaledb/hypertables.md

---

### Distributing a Hypertable across multiple nodes [](distributing)
---

To create a *distributed hypertable* that scales out across many
nodes, execute the TimescaleDB command `create_distributed_hypertable`

> instead of the regular `create_hypertable` command ([API

Reviewer: Suggestion: "Hypertables can be distributed over several nodes using the `create_distributed_hypertable` command instead of the `create_hypertable` command." This will implicitly introduce the notion of a distributed hypertable, so I do not think you need a definition for it.

docs][create_distributed_hypertable]). Note, however, that this

> requires to first configure a distributed database by adding one or

Reviewer 2: ```suggestion
requires you to first configure a distributed database by adding one or
```

> nodes ([API docs][add_data_node]).

Reviewer: Suggestion: "... adding one or more data nodes using [`add_data_node`][add_data_node]."

both the underlying data size *and* any indexes, so some care might be
taken if you make heavy use of expensive index types (e.g., some
PostGIS geospatial indexes).  During testing, you might check your
total chunk sizes via the [chunk relation size][chunk_relation_size]
function.

**Space partitions**: Space partitioning is optional but can make
sense for certain types of data or when using distributed
hypertables. In most cases, space partitioning does not offer any
performance advantages for hypertables that are not distributed.

Space partitions use hashing: Every distinct item is hashed to one of
N buckets. In a distributed hypertable, each bucket corresponds to a
specific data node (although two or more buckets could map to the same
node). Remember that we are already using (flexible) time intervals to
manage chunk sizes; the main purpose of space partitioning is to
enable parallel I/O to the same time interval, both within and across

> nodes.

Reviewer 3: This reads a bit clunky to me, would it be accurate (and clearer) to just say: `the main purpose of space partitioning is to enable parallel I/O within a time interval.`

Author: I actually made significant changes to this section now because I think it is confusing and very out-of-date. See if you like it better now.

Parallel I/O can confer benefits in two scenarios: (a) two or more concurrent
queries should be able to read from different nodes and/or disks in
parallel, and (b) a single query should be able to use query

> parallelization to read from multiple nodes and/or disks in parallel.

Reviewer 3: It's probably also worth mentioning that Parallel I/O can also greatly increase ingest rates when inserting data across several data nodes.

Space partitions use hashing: Every distinct item is hashed to one of
N buckets. In a distributed hypertable, each bucket of the primary
space dimension corresponds to a specific data node (although two or
more buckets could map to the same node). In non-distributed
hypertables, each bucket can map to a distinct disk (using, e.g., a
tablespace).

>:TIP: TimescaleDB does *not* benefit from a very large number of
space partitions (such as the number of unique items you expect in
partition field).  A very large number of such partitions leads both
to poorer per-partition load balancing (the mapping of items to
partitions using hashing), as well as increased planning latency
for some types of queries. We recommend tying the number of space
partitions to the number of disks and/or data nodes.

Spreading chunks along disks and nodes in the space dimension allows
for increased I/O parallelization, either by (a) having multiple
concurrent client processes, or, by (b) splitting the work of a single
client across multiple worker processes on a single node or multiple
concurrent requests across several data nodes.

In summary, to benefit from parallel I/O, on can do one of the
following:

> 1. For each physical disk on a single instance, add a separate

Reviewer 2: Are these steps to be followed, or options?  If the latter, I would use a non-numbered list

tablespace to the database.  TimescaleDB allows you to actually add
multiple tablespaces to a *single* hypertable (although under the
covers, each underlying chunk will be mapped by TimescaleDB to a
single tablespace / physical disk).

> 1. Configure a distributed hypertable that spreads inserts and queries

Reviewer 4: ```suggestion
2. Configure a distributed hypertable that spreads inserts and queries
```
I doubt the number will be automatically updated by markdown due to new line in between.

both the underlying data size *and* any indexes, so some care might be
taken if you make heavy use of expensive index types (e.g., some
PostGIS geospatial indexes).  During testing, you might check your
total chunk sizes via the [chunk relation size][chunk_relation_size]
function.

**Space partitions**: Space partitioning is optional but can make
sense for certain types of data and when using distributed
hypertables.

Space partitions use hashing: Every distinct item is hashed to one of
N buckets. In a distributed hypertable, each bucket of the primary
space dimension corresponds to a specific data node (although two or
more buckets could map to the same node). In non-distributed
hypertables, each bucket can map to a distinct disk (using, e.g., a
tablespace).

>:TIP: TimescaleDB does *not* benefit from a very large number of
space partitions (such as the number of unique items you expect in
partition field).  A very large number of such partitions leads both
to poorer per-partition load balancing (the mapping of items to

> partitions using hashing), as well as much increased planning latency

Reviewer 5: i would remove the `much` here or use different adjective

for some types of queries. We recommend tying the number of space
partitions to the number of disks and/or data nodes.

Spreading chunks along disks and nodes in the space dimension allows
for increased I/O parallelization, either by (a) having multiple
concurrent client processes, or, by (b) splitting the work of a single
client across multiple worker processes on a single node or multiple
concurrent requests across several data nodes.

> Note that query parallelization in PostgreSQL 9.6 (and 10) does

Reviewer 5: this is only true for postgresql Append, ChunkAppend does support query parallelization across chunks even on 9.6 and 10

Author: I guess we can just remove this altogether.

total chunk sizes via the [chunk relation size](#chunk_relation_size)
function.

**Space partitions**: Space partitioning is optional but can make
sense for certain types of data or when using distributed hypertables.

> **In most cases it is not needed for non-distributed hypertables.**

Reviewer: Given the text below about "not benefit from space partitioning", I think you can just replace the boldface sentence with "Space partitioning does not offer any performance advantages for hypertables that are not distributed". Using boldface for anything but very important information or for structuring have a tendency to make the text hard to read.

Space partitions use hashing: Every distinct item is hashed to one of
N buckets. In a distributed hypertable, each bucket corresponds to a
specific data node (although two or more buckets could map to the same
node). Remember that we are already using (flexible) time intervals to
manage chunk sizes; the main purpose of space partitioning is to
enable parallel I/O to the same time interval, both within and across
nodes.

> Parallel I/O can benefit in two scenarios: (a) two or more concurrent

Reviewer 2: ```suggestion
Parallel I/O can confer benefits in two scenarios: (a) two or more concurrent
```

queries should be able to read from different nodes and/or disks in

> parallel, or (b) a single query should be able to use query

Reviewer 2: ```suggestion
parallel, and (b) a single query should be able to use query
```

parallelization to read from multiple nodes and/or disks in parallel.

Note that query parallelization in PostgreSQL 9.6 (and 10) does not

> support querying *different* hypertable chunks in parallel; query

Reviewer 2: Consider turning this into a :warning: element

Author: I am actually unsure this is the case anymore so I'll leave for now. Note also that my changes here are mostly formatting.

parallelization only works on a single physical table (and thus a
single chunk). We might add our own support for this, but it is not
currently supported. Distributed hypertables only work with PostgreSQL

> 11.

Reviewer: Suggest to remove the sentence that we might add it later. This will only pain us.

chunk will be mapped by TimescaleDB to a single tablespace / physical
disk).

1. Configure a distributed hypertable that spreads inserts and queries
across multiple nodes.

> We recommend a RAID setup when possible, as it supports both forms of

Reviewer: s/both forms/all forms/
